src/algokit_utils/_debug_utils.py

```diff
@@ -0,0 +1,160 @@
+                for item in data.get("txn-group-sources", [])
+            ]
+        )
+
+    def to_dict(self) -> dict:
+        return {"txn-group-sources": [json.loads(str(item)) for item in self.txn_group_sources]}
+
+
+@dataclass
+class PersistSourceMapInput:
+    file_name: str
+    teal: str
+    app_name: str
+
+
+def _load_or_create_sources(project_root: Path) -> AVMDebuggerSourceMap:
+    sources_path = project_root / ALGOKIT_DIR / DEBUGGER_DIR / SOURCES_FILE
+    if not sources_path.exists():
+        return AVMDebuggerSourceMap(txn_group_sources=[])
+
+    with sources_path.open() as f:
+        return AVMDebuggerSourceMap.from_dict(json.load(f))
+
+
+def _upsert_debug_sourcemaps(sourcemaps: list[AVMDebuggerSourceMapEntry], project_root: Path) -> None:
+    sources_path = project_root / ALGOKIT_DIR / DEBUGGER_DIR / SOURCES_FILE
+    sources = _load_or_create_sources(sources_path)
+
+    for sourcemap in sourcemaps:
+        if sourcemap not in sources.txn_group_sources:
+            sources.txn_group_sources.append(sourcemap)
+
+    with sources_path.open("w") as f:
+        json.dump(sources.to_dict(), f)
+
+
+def _build_avm_sourcemap(
+    teal_content: str, app_name: str, file_name: str, output_path: Path, client: "AlgodClient"
+) -> AVMDebuggerSourceMapEntry:
+    file_name = f"{file_name}.teal" if not file_name.endswith(".teal") else file_name
+    result = client.compile(deploy.strip_comments(teal_content), source_map=True)
+    program_hash = base64.b64encode(
+        checksum(base64.b64decode(result["result"]))  # type: ignore[no-untyped-call]
+    ).decode()
+    source_map = SourceMap(result["sourcemap"]).__dict__
+    source_map["sources"] = [file_name]
+
+    output_dir = output_path / ALGOKIT_DIR / DEBUGGER_DIR / app_name
+    source_map_output = output_dir / f'{file_name.replace(".teal", "")}.tok.map'
+    source_map_output.parent.mkdir(parents=True, exist_ok=True)
+    source_map_output.write_text(json.dumps(source_map))
+    (output_dir / file_name).write_text(teal_content)
+
+    return AVMDebuggerSourceMapEntry(str(source_map_output), program_hash)
+
+
+def persist_sourcemaps(
+    sources: list[PersistSourceMapInput],
+    project_root: Path,
+    client: "AlgodClient",
+) -> None:
+    sourcemaps = [
+        _build_avm_sourcemap(source.teal, source.app_name, source.file_name, project_root, client) for source in sources
+    ]
+
+    _upsert_debug_sourcemaps(sourcemaps, project_root)
+
+
+def simulate_response(atc: AtomicTransactionComposer, algod_client: "AlgodClient") -> SimulateAtomicTransactionResponse:
+    unsigned_txn_groups = atc.build_group()
+    empty_signer = EmptySigner()
+    txn_list = [txn_group.txn for txn_group in unsigned_txn_groups]
+    fake_signed_transactions = empty_signer.sign_transactions(txn_list, [])
+    txn_group = [SimulateRequestTransactionGroup(txns=fake_signed_transactions)]
+    trace_config = SimulateTraceConfig(enable=True, stack_change=True, scratch_change=True)
```

Comment:
There is also a `state_change` property that I think we want to enable.

Reply:
@neilcampbell where can i find that property?
![CleanShot 2023-11-23 at 16 13 26](https://github.com/algorandfoundation/algokit-utils-py/assets/7698600/89cd7b56-41a2-4f5e-a6d7-b63bd1038b7b)
those seem to be the only options in simulate config object

Reply:
Interesting! I noticed it on the JS sdk the other day and assumed it'd be on the python sdk.
https://algorand.github.io/js-algorand-sdk/classes/modelsv2.SimulateTraceConfig.html

Reply:
It's also configurable on the algod request https://developer.algorand.org/docs/rest-apis/algod/#simulatetraceconfig

Reply:
@neilcampbell i think feature parity issue? Here is an issue to track on js sdk repo https://github.com/algorand/py-algorand-sdk/issues/519